wlm/src/main/java/ai/djl/serving/wlm/ModelInfo.java

```diff
@@ -588,6 +609,143 @@ private String inferEngine() throws ModelNotFoundException {
         throw new ModelNotFoundException("Failed to detect engine of the model: " + modelDir);
     }
 
+    private boolean isPythonModel(String prefix) {
+        return Files.isDirectory(modelDir.resolve("MAR-INF"))
+                || Files.isRegularFile(modelDir.resolve("model.py"))
+                || Files.isRegularFile(modelDir.resolve(prefix + ".py"))
+                || Utils.getEnvOrSystemProperty("HF_MODEL_ID") != null
+                || Files.isRegularFile(modelDir.resolve("config.json"))
+                || prop.containsKey("option.s3url")
+                || prop.containsKey("option.model_id");
+    }
+
+    private String inferLMIEngine() throws ModelException, IOException {
+        // MMS/Torchserve
+        if (Files.isDirectory(modelDir.resolve("MAR-INF"))) {
+            return "Python";
+        }
+        JsonObject modelConfig = getHuggingFaceModelConfig();
+        if (modelConfig == null) {
+            return "Python";
+        }
+
+        String modelType;
+        if (modelConfig.has("_diffusers_version")) {
+            modelType = "stable-diffusion";
+        } else {
+            modelType = modelConfig.get("model_type").getAsString();
+        }
+        long numAttentionHeads = Long.MAX_VALUE;
+        // All of these are valid in the config.json file for the number of attention heads
+        if (modelConfig.has("num_attention_heads")) {
+            numAttentionHeads = modelConfig.get("num_attention_heads").getAsLong();
+        } else if (modelConfig.has("num_heads")) {
+            numAttentionHeads = modelConfig.get("num_heads").getAsLong();
+        } else if (modelConfig.has("n_head")) {
+            numAttentionHeads = modelConfig.get("n_head").getAsLong();
+        }
+        int tensorParallelDegree = CudaUtils.getGpuCount();
+        if (Utils.getEnvOrSystemProperty("TENSOR_PARALLEL_DEGREE") != null) {
+            tensorParallelDegree =
+                    Integer.parseInt(Utils.getEnvOrSystemProperty("TENSOR_PARALLEL_DEGREE"));
+        } else if (prop.get("option.tensor_parallel_degree") != null) {
+            tensorParallelDegree =
+                    Integer.parseInt(prop.get("option.tensor_parallel_degree").toString());
+        }
+
+        if (tensorParallelDegree > 0) {
+            prop.put("option.tensor_parallel_degree", tensorParallelDegree);
+        }
+        String huggingFaceTask = Utils.getEnvOrSystemProperty("HF_TASK");
+        if (huggingFaceTask != null) {
+            prop.put("option.task", huggingFaceTask);
+        }
+
+        if ("stable-diffusion".equals(modelType)) {
+            // TODO: Move this from hardcoded to deduced in PyModel
+            prop.put("option.entryPoint", "djl_python.stable-diffusion");
+            return "DeepSpeed";
+        }
+
+        if (!isTensorParallelSupported(numAttentionHeads, tensorParallelDegree)) {
```

Comment:
Maybe DS or FT would have some mechanism from their end to decide how to do model sharding. I would suggest to not check this

Reply:
At least for DS, they will throw an exception if this check fails. But really the only practical examples of this we have seen is gpt2-xl.

In the future it's possible that DS and FT change that behavior and can actually accommodate such a model. At that point this method would become incorrect.

I can remove this, since it's going to be validated by the engine anyways. But the benefit of doing it this way is that we don't recommend say gpt2-xl to run with DeepSpeed with TP when we know it won't work.